libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,376 @@
+	{ ops.pushOrDupTarget(targetOffset) };
+};
+*/
+/// Helper class that can perform shuffling of a source stack layout to a target stack layout via
+/// abstracted shuffle operations.
+template</*ShuffleOperationConcept*/ typename ShuffleOperations>
+class Shuffler
+{
+public:
+	/// Executes the stack shuffling operations. Instantiates an instance of ShuffleOperations
+	/// in each iteration. Each iteration performs exactly one operation that modifies the stack.
+	/// After shuffle, source and target have the same size and all slots in the source layout are
+	/// compatible to the slots at the same target offset.
+	template<typename... Args>
+	static void shuffle(Args&&... args)
+	{
+		bool needsMoreShuffling = true;
+		// The shuffling algorithm should always terminate in polynomial time, but we provide a limit
+		// in case it does not terminate due to a bug.
+		size_t iterationCount = 0;
+		while (iterationCount < 1000 && (needsMoreShuffling = shuffleStep(std::forward<Args>(args)...)))
+			++iterationCount;
+		yulAssert(!needsMoreShuffling, "Could not create stack layout after 1000 iterations.");
+	}
+private:
+	// If dupping an ideal slot causes a slot that will still be required to become unreachable, then dup
+	// the latter slot first.
+	// @returns true, if it performed a dup.
+	static bool dupDeepSlotIfRequired(ShuffleOperations& _ops)
+	{
+		// Check if the stack is large enough for anything to potentially become unreachable.
+		if (_ops.sourceSize() < 15)
+			return false;
+		// Check whether any deep slot might still be needed later.
+		for (size_t sourceOffset: ranges::views::iota(0u, _ops.sourceSize() - 15))
+		{
+			// We need another copy of this slot.
+			if (_ops.sourceMultiplicity(sourceOffset) > 0)
+			{
+				// If this slot occurs again later, we skip this occurrence.
+				if (ranges::any_of(
+					ranges::views::iota(sourceOffset + 1, _ops.sourceSize()),
+					[&](size_t _offset) { return _ops.sourceIsSame(sourceOffset, _offset); }
+				))
+					continue;
+				// Bring up the target slot that would otherwise become unreachable.
+				for (size_t targetOffset: ranges::views::iota(0u, _ops.targetSize()))
+					if (!_ops.targetIsArbitrary(targetOffset) && _ops.isCompatible(sourceOffset, targetOffset))
+					{
+						_ops.pushOrDupTarget(targetOffset);
+						return true;
+					}
+			}
+		}
+		return false;
+	}
+	static void bringUpTargetSlot(ShuffleOperations& _ops, size_t _targetOffset)
+	{
+		std::list<size_t> toVisit{_targetOffset};
+		std::set<size_t> visited;
+
+		while (!toVisit.empty())
+		{
+			auto offset = *toVisit.begin();
+			toVisit.erase(toVisit.begin());
+			visited.emplace(offset);
+			if (_ops.targetMultiplicity(offset) > 0)
+			{
+				_ops.pushOrDupTarget(offset);
+				return;
+			}
+			// The desired target slot must already be somewhere else on stack right now.
+			for (auto nextOffset: ranges::views::iota(0u, std::min(_ops.sourceSize(), _ops.targetSize())))
```

Comment:
What if `0` is too deep in the stack? Shouldn't you search starting from the top?

Reply:
This is maybe lacking some documentation... and the function is badly named. And it's probably the most complicated case in the entire shuffling.
What this function does is not generally pushing the slot that is supposed to end up at ``_targetOffset`` of the target. That only happens, if I actually need another copy of that slot, in which case the very first iteration will have ``_ops.targetMultiplicity(offset) > 0`` and immediately push or dup the slot.

The situation in which this is called is that I want to *fix* ``target[_targetOffset]`` next and I want to dup *something* that will lead to that being fixed down the road. Pushing ``target[_targetOffset]`` itself is just the simplest case, in which there aren't already enough copies of that particular slot on the current stack.

If there are already enough copies of ``target[_targetOffset]`` itself on stack, then I must be able to find an occurrence of that slot in the *current source* that is not yet in place. That's what the inner loop is looking for, i.e. that slot ends up being ``nextOffset``. Now that slot not yet being in place means that the *target* slot at ``nextOffset`` should be fixed next. If I can dup that target slot, i.e. ``multiplicity[nextOffset] > 0``, then I can dup it, the next iteration will swap it into place and *then* I have the slot that was supposed to end up at ``_targetOffset`` at the stack top and can fix ``_targetOffset`` by another swap.

But there's no guarantee that I don't already have enough copies of ``nextOffset`` on the current stack as well, i.e. I'm recursively in the same situation as before: I want to fix ``nextOffset`` by dupping *something*, just as before I wanted to fix ``_targetOffset`` by dupping *something*.

Eventually, I have to hit something that I *can* dup and then there will be a cascading effect fixing all the target slots I visited via swapping.

Or in other words: the situation is that there is be an arbitrary long cycle starting at ``_targetOffset`` that can be broken by dupping *something* and then have swapping take care of permuting the entire cycle. The simplest case is a cycle of length 1, i.e. I push ``target[_targetOffset]`` itself. But it can be arbitrarily long cycles.

Before looking at what gasol does, I actually thought that there was no optimal choice for dupping in cases like this and just blindly dupped the first thing with ``multiplicity > 0`` - only after thinking through what gasol did on those cases I realized that there will always be a cycle and thereby an optimal choice :-).

Reply:
Can you condense this into 1-2 sentences above the function?

Reply:
Also - what if `0` is too deep in the stack?

Reply:
If 0 is too deep in the stack, i.e. the bottom most stack slot is not in finalized, but I cannot reach it, then there's little we can do anyways, is there? At least pushing any other slot will only make things worse, so better to report this very slot as unreachable right away.

One thing one *could* try to do and what is not done yet, if a too deep slot is not finalized yet, but already unreachable, is to check if it's possible to compress the stack, i.e. pop all duplicates and all slots that can be freely generated first, until hopefully the slot in question is reachable again. Not sure if we should add that right now or leave it as future improvement option.

But in general this entire algorithm is only guaranteed to succeeed, if unlimited stack access is possible. It is always allowed to just request to dup a too deep slot and then the caller will register that as a stack too deep error.